Update basic/stl-sort (#23)
OI-wiki/i18n: docs/basic/stl-sort.md

## sort

The standard C library implements quick sort, which is `qsort` in `stdlib.h`.

But in OI-related competitions, the library sorting function more commonly used is the `std::sort` from the C++ `algorithm` library.

> The C++ standard does not strictly require the algorithm implemented of this function; the specific implementation depends on the compiler.

Reviewer: specified the implementation of this function and it depends on the compiler (used?) 即可

The old version of the C++ standard only requires its **average** time complexity to reach $O(n\log n)$ , but the C++11 standard requires its **worst** time complexity to reach $O( n\log n)$ . You can refer to [std::sort()](https://en.cppreference.com/w/cpp/algorithm/sort).

在 [libstdc++](https://github.com/mirrors/gcc/blob/master/libstdc++-v3/include/bits/stl_algo.h) 和 [libc++](http://llvm.org/svn/llvm-project/libcxx/trunk/include/algorithm) 中使用的都是 [Introsort](https://en.wikipedia.org/wiki/Introsort) 。

> In both s[libstdc++](https://github.com/mirrors/gcc/blob/master/libstdc++-v3/include/bits/stl_algo.h) and [libc++](http://llvm.org/svn/llvm-project/libcxx/trunk/include/algorithm), [introsort](https://en.wikipedia.org/wiki/Introsort) is used.

Reviewer: ```suggestion
In both [libstdc++](https://github.com/mirrors/gcc/blob/master/libstdc++-v3/include/bits/stl_algo.h) and [libc++](http://llvm.org/svn/llvm-project/libcxx/trunk/include/algorithm), [introsort](https://en.wikipedia.org/wiki/Introsort) is used.
```

Introsort limits the divide and conquer depth of quicksort. When divide and conquer reaches a certain depth, use a sorting algorithm (such as heap sort) with the worst time complexity of $O(n\log n)$ to sort the sub-arrays.

This limitation of Introsort makes its worst time complexity $O(n\log n)$ .

Implementation:

> 

Reviewer: Usage

刚刚没看中文还以为原文连 introsort 的实现都写了 🌚

Reviewer: 下面也要改一下

## partial_sort

> Place the top $k$ smallest elements in the first $k$ positions in order in the sequence, and the order following that is not guaranteed.

Reviewer: 根据 [cppreference](https://en.cppreference.com/w/cpp/algorithm/partial_sort) 的说明来看，好像还是排序了前 k 个（从函数名也可以看出来），这样翻译有点像相对顺序不变了。

Author: 加了个注释，不知道这样行不行 :thinking:

> Place the top $k$ smallest elements in the first $k$ positions in the sequence (the order of equal of elements is not guarnteed), and the order of the remaining elements is not specified.

Reviewer: ```suggestion
Place the top $k$ smallest elements in the first $k$ positions in the sequence (the order of equal elements is not guarnteed), and the order of the remaining elements is not specified.
```

Time complexity: $O(n\log k)$ 

Implementation:

```cpp
std::partial_sort(begin, begin + k, end);
std::partial_sort(begin, begin + k, end, cmp);
```

Principle:

> The principle of implementing `partial_sort` is: perform the `make_heap()` operation on the elements from the interval of $[first, middle)$ in the original sequence to construct a max heap, and then take each element in $[middle, last)$ to compare with the $first$ - the element in $first$ is the maximum value in the heap. If it is less than the maximum value, the element positions are swapped, and the elements in $[first, middle)$ are adjusted to maintain the max heap. After the comparison, the elements in $[first, middle)$ are sorted by `sort_heap()` in ascending order. Please note that heap order and ascending order are different.

Reviewer: ```suggestion
The principle of implementing `partial_sort` is to perform the `make_heap()` operation on the elements in the interval $[first, middle)$ in the original sequence to construct a max heap, and then take each element in $[middle, last)$ to compare with the $first$ — the maximum value in the heap. If it is less than the maximum value, the two elements are swapped, and the elements in $[first, middle)$ are adjusted to maintain the max heap. After the comparison, the elements in $[first, middle)$ are sorted by `sort_heap()` in ascending order. Please note that heap order is different ascending order.
```

## Define operator

> For both built-in types (such as `int`) and user-defined structures, you can define the **less than operator** when calling the STL sorting function. You can pass a comparison operator function (usually the last item) when calling the function, or directly override the default operator of this type. See [cppreference](https://en.cppreference.com/w/cpp/language/operators).

Reviewer: ```suggestion
For both built-in types (such as `int`) and user-defined structures, you can define the **less than operator** when calling the STL sorting function. You can pass a comparison operator function (usually the last parameter) when calling the function, or directly override the default operator of this type. See [cppreference](https://en.cppreference.com/w/cpp/language/operators).
```

  return (u1.a == u2.a) ? (u1.b > u2.b) : (u1.a > u2.a);
}
// ......
std::sort(da + 1, da + 1 + 10);  // use the <operator defined in the structure to sort ascendingly
std::sort(da + 1, da + 1 + 10, cmp);  // use cmp function to compare, sort descendingly
```

### Strictly weak order

The sorting operator must satisfy the [strict weak ordering](https://en.wikipedia.org/wiki/Weak_ordering#:~:text=A%20strict%20weak%20ordering%20is,b%20%3C%20a%22%20is%20transitive.&text=For%20all%20x%2C%20y%2C%20z,x%20%3C%20z%20(transitivity).), otherwise there will be unpredictable situations (such as runtime error, can not be sorted correctly, etc).

The requirements of the strict weak order:

1. $x \not< x$ (non-reflexive)
2. If $x < y$ , then $y \not< x$ (asymmetry)
3. If $x < y, y < z$ , then $x < z$ (transitivity)

> 4. If $x \not< y, y \not< x, y \not< z, z \not< y$ , then $x \not< z, z \not< x$ (incomparable transitivity )

Reviewer: ```suggestion
4. If $x \not< y, y \not< x, y \not< z, z \not< y$ , then $x \not< z, z \not< x$ (incomparable transitivity)
```
